### DISCUSSION TITLE
Unable to accomplish throttling

### DISCUSSION THREAD

**Discussion by suchigg** [vladimir-bukhtoyarov/bucket4j#299] [Q&A] [answered]
Hello, 
I tried to follow documentation to implement an API call throttling as described [here](https://bucket4j.com/8.1.0/toc.html#using-bucket-as-throttler) but without any success. 
I found in an old thread that "Token-bucket math model is not addressed to satisfy hard limits like N permit /M time. Please read this discussion #128 " but this looks in contrary to the documentation description.
Thanks in advance

**Comment by vladimir-bukhtoyarov** [comment]
Hello @suchigg 

Please provide the code(and comments) that show the problem.

**Reply by suchigg** [reply]
Hello Vladimir and thank you for your support
This is the code I wrote to make some tests. 

`public class BucketTest {

	public static void main(String[] args) throws InterruptedException {
		Bandwidth operationalLimit = Bandwidth.simple(30, Duration.ofSeconds(30));
		Bucket bucket = Bucket.builder()
							.addLimit(operationalLimit)
							.build();

		while (true) {
			System.out.printf("available tokens: %d \n", bucket.getAvailableTokens());
			bucket.asScheduler().consume(1, Executors.newScheduledThreadPool(1));
			doSomething();
		}
	}

	private static void doSomething() throws InterruptedException {
		Thread.sleep(250);
		System.out.println("doSomething Method Called at: " + LocalDateTime.now());
	}

}`

this is the output:

available tokens: 30 
doSomething Method Called at: 2022-09-01T09:20:42.938936300
available tokens: 29 
doSomething Method Called at: 2022-09-01T09:20:43.221216200
available tokens: 28 
doSomething Method Called at: 2022-09-01T09:20:43.476211100
available tokens: 27 
doSomething Method Called at: 2022-09-01T09:20:43.728993200
available tokens: 27 
doSomething Method Called at: 2022-09-01T09:20:43.981990900
available tokens: 26 
doSomething Method Called at: 2022-09-01T09:20:44.234343500
available tokens: 25 
doSomething Method Called at: 2022-09-01T09:20:44.484530100
available tokens: 24 
doSomething Method Called at: 2022-09-01T09:20:44.735714200
available tokens: 24 
doSomething Method Called at: 2022-09-01T09:20:45.000307200
available tokens: 23 
doSomething Method Called at: 2022-09-01T09:20:45.252219600
available tokens: 22 
doSomething Method Called at: 2022-09-01T09:20:45.503864400
available tokens: 21 
doSomething Method Called at: 2022-09-01T09:20:45.767503400
available tokens: 21 
doSomething Method Called at: 2022-09-01T09:20:46.032356800
available tokens: 20 
doSomething Method Called at: 2022-09-01T09:20:46.295146300
available tokens: 19 
doSomething Method Called at: 2022-09-01T09:20:46.545747500
available tokens: 18 
doSomething Method Called at: 2022-09-01T09:20:46.798313900
available tokens: 18 
doSomething Method Called at: 2022-09-01T09:20:47.063313200
available tokens: 17 
doSomething Method Called at: 2022-09-01T09:20:47.327592500
available tokens: 16 
doSomething Method Called at: 2022-09-01T09:20:47.591051500
available tokens: 15 
doSomething Method Called at: 2022-09-01T09:20:47.856183900
available tokens: 15 
doSomething Method Called at: 2022-09-01T09:20:48.120618
available tokens: 14 
doSomething Method Called at: 2022-09-01T09:20:48.373246100
available tokens: 13 
doSomething Method Called at: 2022-09-01T09:20:48.638287800
available tokens: 12 
doSomething Method Called at: 2022-09-01T09:20:48.890334400
available tokens: 12 
doSomething Method Called at: 2022-09-01T09:20:49.141428200
available tokens: 11 
doSomething Method Called at: 2022-09-01T09:20:49.391572400
available tokens: 10 
doSomething Method Called at: 2022-09-01T09:20:49.646186700
available tokens: 9 
doSomething Method Called at: 2022-09-01T09:20:49.911499800
available tokens: 9 
doSomething Method Called at: 2022-09-01T09:20:50.163498900
available tokens: 8 
doSomething Method Called at: 2022-09-01T09:20:50.428686900
available tokens: 7 
doSomething Method Called at: 2022-09-01T09:20:50.681012300
available tokens: 7 
doSomething Method Called at: 2022-09-01T09:20:50.932664700
available tokens: 6 
doSomething Method Called at: 2022-09-01T09:20:51.197654500
available tokens: 5 
doSomething Method Called at: 2022-09-01T09:20:51.461185500
available tokens: 4 
doSomething Method Called at: 2022-09-01T09:20:51.711748800
available tokens: 4 
doSomething Method Called at: 2022-09-01T09:20:51.976455300
available tokens: 3 
doSomething Method Called at: 2022-09-01T09:20:52.242367200
available tokens: 2 
doSomething Method Called at: 2022-09-01T09:20:52.506275
available tokens: 1 
doSomething Method Called at: 2022-09-01T09:20:52.771202500
available tokens: 1 
doSomething Method Called at: 2022-09-01T09:20:53.039756400
available tokens: 0 
doSomething Method Called at: 2022-09-01T09:20:53.303001300
available tokens: -1 
doSomething Method Called at: 2022-09-01T09:20:53.567111100
available tokens: -2 
doSomething Method Called at: 2022-09-01T09:20:53.832249600
available tokens: -2 
doSomething Method Called at: 2022-09-01T09:20:54.083250300
available tokens: -3 
doSomething Method Called at: 2022-09-01T09:20:54.349360700
available tokens: -4 
doSomething Method Called at: 2022-09-01T09:20:54.601604
available tokens: -5 
doSomething Method Called at: 2022-09-01T09:20:54.866427300
available tokens: -5 
doSomething Method Called at: 2022-09-01T09:20:55.118027400
available tokens: -6 
doSomething Method Called at: 2022-09-01T09:20:55.369084800
available tokens: -7 
doSomething Method Called at: 2022-09-01T09:20:55.634133900
available tokens: -8 
doSomething Method Called at: 2022-09-01T09:20:55.897683200
available tokens: -8 
doSomething Method Called at: 2022-09-01T09:20:56.163242300
available tokens: -9 
doSomething Method Called at: 2022-09-01T09:20:56.429173400
available tokens: -10 
doSomething Method Called at: 2022-09-01T09:20:56.694414100
available tokens: -10

The expected behavior, based on the documentation i linked, is that it's not possible to have negative numbers of tokens.
I tried to change Bandwith as `Bandwidth.classic(30, Refill.intervally(30, Duration.ofSeconds(30)));` and make some other attempt using `ConsumptionProbe` to obtain nanos to reset and do some math, but without success.

Thank you in advance

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
It is bug in documentation. I will fix it.  You need to use ```asBlocking``` instead of  ```asScheduler``` in case of you need to block current thread until requested amount of tokens become available.

**Reply by suchigg** [reply]
Thank you for the reply.
Another question. is there a way to start token regeneration only after the first tryConsume call and reset if after a threshold time? 
If i understand how the library works, refill time of bandwidth is independent of token consumption, so i think i need to create some type of mechanism to start the bucket when is used for the first time and reset it after a certain time has passed

**Reply by vladimir-bukhtoyarov** [reply]
> is there a way to start token regeneration only after the first tryConsume

No, there is no such functionality.

> and reset if after a threshold time

The same. There is no such functionality

**Reply by suchigg** [reply]
Thank you again for the answer, i suppose i need to implement something custom then, unfortunately.
Anyway, your library looks amazing, good job!

Kind regards,
Luca
